src/lib/circuit.ts

```diff
@@ -0,0 +1,221 @@
+  SnarkyVerificationKey,
+} from '../snarky.js';
+import { SnarkyCircuit } from './circuit_value.js';
+import { withThreadPool } from '../snarkyjs-bindings/js/wrapper.js';
+
+// external API
+export { public_, circuitMain, Circuit };
+
+class Circuit {
+  // circuit-writing interface
+
+  /**
+   * Generates a proving key and a verification key for this circuit.
+   */
+  static generateKeypair() {
+    return withThreadPool(async () =>
+      SnarkyCircuit.generateKeypair(this as any)
+    );
+  }
+
+  /**
+   * Proves a statement using the private input, public input and the {@link Keypair} of the circuit.
+   */
+  static prove(privateInput: any[], publicInput: any[], keypair: Keypair) {
+    return withThreadPool(async () =>
+      SnarkyCircuit.prove(this as any, privateInput, publicInput, keypair)
+    );
+  }
+
+  /**
+   * Verifies a proof using the public input, the proof and the initial {@link Keypair} of the circuit.
+   */
+  static verify(
+    publicInput: any[],
+    vk: SnarkyVerificationKey,
+    proof: SnarkyProof
+  ) {
+    return withThreadPool(async () =>
+      SnarkyCircuit.verify(publicInput, vk, proof)
+    );
+  }
+
+  // utility namespace
+
+  /**
+   * Create a new witness. A witness, or variable, is a value that is provided as input
+   * by the prover. This provides a flexible way to introduce values from outside into the circuit.
+   * However, note that nothing about how the value was created is part of the proof - `Circuit.witness`
+   * behaves exactly like user input. So, make sure that after receiving the witness you make any assertions
+   * that you want to associate with it.
+   *
+   * Example for re-implementing `Field.inv` with the help of `witness`:
+   * ```
+   * let invX = Circuit.witness(Field, () => {
+   *   // compute the inverse of `x` outside the circuit, however you like!
+   *   return Field.inv(x));
+   * }
+   * // prove that `invX` is really the inverse of `x`:
+   * invX.mul(x).assertEquals(1);
+   * ```
+   */
+  static witness = SnarkyCircuit.witness;
+  /**
+   * Runs code as a prover.
+   */
+  static asProver = SnarkyCircuit.asProver;
+  /**
+   * Runs code and checks its correctness.
+   */
+  static runAndCheck = SnarkyCircuit.runAndCheck;
+  /**
+   * Returns information about the constraint system in the callback function.
+   */
+  static constraintSystem = SnarkyCircuit.constraintSystem;
+  /**
+   * Returns a low-level JSON representation of the `Circuit` from its {@link Keypair}:
+   * a list of gates, each of which represents a row in a table, with certain coefficients and wires to other (row, column) pairs
+   */
+  static constraintSystemFromKeypair =
+    SnarkyCircuit.constraintSystemFromKeypair;
+  /**
+   * Creates a {@link Provable} for a generic array.
+   */
+  static array = SnarkyCircuit.array;
+  /**
+   * Asserts that two values are equal.
+   */
+  static assertEqual = SnarkyCircuit.assertEqual;
+  /**
+   * Checks if two elements are equal.
+   */
```

Comment:
I'm going to change this so that it shows how to use for Structs (for which it is most useful -- on `Field` we already have `.equals()`)

```suggestion
  /**
   * Checks if two elements are equal.
   * @example
   * ```ts
   * class MyStruct extends Struct({ a: Field, b: Bool }) {};
   * const a: MyStruct = { a: Field(0), b: Bool(false) };
   * const b: MyStruct = { a: Field(1), b: Bool(true) };
   * const isEqual = Circuit.equal(MyStruct, a, b);
   */
```